### PR TITLE
Making tq stats visible from cli

### PR DESCRIPTION
**Pull request by Shivs11**
## What was changed
- Adds support to view TQ stats from the CLI.
- Sample input and output for the command:

`temporal task-queue describe --task-queue xxx`

### Task Queue Statistics

| BuildID     | TaskQueueType | ApproximateBacklogCount | ApproximateBacklogAge | BacklogIncreaseRate | TasksAddRate | TasksDispatchRate |
|-------------|---------------|-------------------------|-----------------------|---------------------|--------------|-------------------|
| UNVERSIONED | workflow       | 0                       | 0s                    | 0                   | 0            | 0                 |
| UNVERSIONED | activity       | 0                       | 0s                    | 0                   | 0            | 0                 |

`temporal task-queue describe --task-queue xxx  -o json`

`
"stats": [
  {
    "buildId": "UNVERSIONED",
    "taskQueueType": "workflow",
    "approximateBacklogCount": 0,
    "approximateBacklogAge": 0,
    "backlogIncreaseRate": 0,
    "tasksAddRate": 0,
    "tasksDispatchRate": 0
  },
  {
    "buildId": "UNVERSIONED",
    "taskQueueType": "activity",
    "approximateBacklogCount": 0,
    "approximateBacklogAge": 0,
    "backlogIncreaseRate": 0,
    "tasksAddRate": 0,
    "tasksDispatchRate": 0
  }
]
`

## Why?
- Shall help in worker fleet management.

## Checklist

1. Closes 

2. How was this tested:
- Modified an existing unit test
- Ensured all other unit tests pass

3. Any docs updates needed?

- Updated docs.

### CODE REVIEW

File: go.mod
Comment (Shivs11) [inline go.mod:18]: Upgraded sdk to point to the version where TQ stats were introduced in DescribeTaskQueueEnhanced (#1553)

File: temporalcli/commands.taskqueue.go
Comment (ShahabT) [inline temporalcli/commands.taskqueue.go:166]: Nit: should have empty line between functions

File: temporalcli/commands.taskqueue_test.go
Comment (ShahabT) [inline temporalcli/commands.taskqueue_test.go:91]: can we have a test with non-zero results?

File: temporalcli/commands.gen.go
Comment (ShahabT) [inline temporalcli/commands.gen.go:1560]: what's the default value? I'm thinking it should be true by default.

Comment (Shivs11) [reply temporalcli/commands.gen.go:1560]: hmm, why do you think it should be true? The default value is false and I think it should be kept that way - noticed other flags were also following a similar principle unless we have something specific for this one

**Review by cretz — COMMENTED** [review]
Can you show what the output of this command w/ stats for a normal task queue (no versioning) looks like in text and JSON form? Can just be as a comment here in the PR. It's hard to get a feel for what the user sees from the code alone.

File: temporalcli/commands.gen.go
Comment (Shivs11) [inline temporalcli/commands.gen.go:1560]: @ShahabT  - Since this has now been changed to `disable-stats` with a default value of false, it brings us back to the bug we found in matching where a versioned queue may not be loaded leading to null dereferences. Again, not a big deal as you said since versioning happens to be pre-release but just wanted to make it a note here.

File: temporalcli/commands.taskqueue_test.go
Comment (ShahabT) [inline temporalcli/commands.taskqueue_test.go:84]: `NotContains` would return true even if nothing is printed. I think we should check the things we want to be shown are shown. (if backlog age is an issue, we should skip that somehow or put a wildcard there).

Also, we should put these checks in EventuallyWithT because the stats may not be updated right away and that'd make the test flaky.

Lastly, I'd check for stats of both wf and activity types, even if activity is expected to have zero stats. This would catch the case if by accident (now or in the future) we print workflow stats for activity as well.

File: temporalcli/commands.taskqueue.go
Comment (ShahabT) [inline temporalcli/commands.taskqueue.go:45]: What's the benefit of separate maps for build ID -> reachability, build ID -> poller, and build ID -> stats? isn't better to have a single map of build ID -> {reachability, pollers, stats}? cc @antlai-temporal @cretz

Comment (Shivs11) [reply temporalcli/commands.taskqueue.go:45]: I too, am not totally opposed to this idea - included the new field cause I thought a potential norm could have been to have separate maps for each but would love to hear your thoughts @cretz

File: temporalcli/commands.gen.go
Comment (Shivs11) [reply temporalcli/commands.gen.go:1560]: Outdated comment since a patch was merged to main last week - [PR](https://github.com/temporalio/temporal/pull/6437)

File: temporalcli/commands.taskqueue.go
Comment (cretz) [reply temporalcli/commands.taskqueue.go:45]: I have no strong opinions on the structure of the output (nor did I when this was first developed). Anything sane works for me. But just remember, it must be stable.

**Comment by ShahabT** [comment]
@antlai-temporal @Shivs11 I think it's better to change the short and long descriptions of the `describe` command. Right now it seems we're over-focusing on the reachability info. e.g. the short description is "Provides task reachability and pollers information for Workers on this Task Queue.".

I'd suggest we mention poller info and stats as the main output. Reachability can be mentioned as a experimental part of the API with proper warning that it may change in the future. Note that there is a chance that we want to move reachability API under `temporal versioning` command, and in that case it's not clear if we'd still keep it under `temporal task-queue describe` too.

**Comment by Shivs11** [comment]
> @antlai-temporal @Shivs11 I think it's better to change the short and long descriptions of the `describe` command. Right now it seems we're over-focusing on the reachability info. e.g. the short description is "Provides task reachability and pollers information for Workers on this Task Queue.".
> 
> I'd suggest we mention poller info and stats as the main output. Reachability can be mentioned as a experimental part of the API with proper warning that it may change in the future. Note that there is a chance that we want to move reachability API under `temporal versioning` command, and in that case it's not clear if we'd still keep it under `temporal task-queue describe` too.

I was not aware that the reachability side of things could be moved over later and knowing this now, it makes sense to not over-focus on it. I am for the act of keeping pollers and stats as the main output but shall wait for @antlai-temporal to give us the green light 🚥

File: temporalcli/commands.taskqueue.go
Comment (ShahabT) [inline temporalcli/commands.taskqueue.go:36]: What's the type of age in json? is it a number? if so, we need to qualify the name with something as "approximateBacklogAgeSeconds".

Comment (Shivs11) [reply temporalcli/commands.taskqueue.go:36]: We are returning a Duration (effectively, nanoseconds) which gets printed for JSON. Yes, it is a number. Sample output - 

``` "approximateBacklogAge": 4108891000, ```

Comment (Shivs11) [reply temporalcli/commands.taskqueue.go:36]: As discussed offline, the type for JSON was changed from `time.Duration` to `string`. Thus, I think it's okay to keep the same name.

**Comment by cretz** [comment]
> Can you show what the output of this command w/ stats for a normal task queue (no versioning) looks like in text and JSON form?

I saw the markdown table in PR description but I think we want to see what the raw table looks like in text (and will wait until after y'all's discussion on text/json format to review what latest output looks like and approve).

**Comment by Shivs11** [comment]
> > Can you show what the output of this command w/ stats for a normal task queue (no versioning) looks like in text and JSON form?
> 
> I saw the markdown table in PR description but I think we want to see what the raw table looks like in text (and will wait until after y'all's discussion on text/json format to review what latest output looks like and approve).

Ah, my apologies for the misunderstanding. Pasting the output as seen from the CLI here:

Scenario - Adding a task in the backlog with no pollers:

1. JSON 
image

2. Non-JSON
image

note - the backlog increase rate becomes 0 in the second picture because we are fetching the rate over a window of 30 seconds which were elapsed between the two screenshots.

**Review by cretz — APPROVED** [review]
LGTM, but would like to wait for other approvers because I think they may have stronger opinions on the output structure. And please don't merge until we have a tagged SDK dependency.

File: temporalcli/commands.taskqueue.go
Comment (antlai-temporal) [reply temporalcli/commands.taskqueue.go:45]: It is just for modularity, these are very different top-level concepts let's keep them separate. The overhead of three maps is just minimal...

**Comment by antlai-temporal** [comment]
> I was not aware that the reachability side of things could be moved over later and knowing this now, it makes sense to not 
> over-focus on it. I am for the act of keeping pollers and stats as the main output but shall wait for @antlai-temporal to give us the green light

The main reason to over focus on reachability was the feedback that it was hard to find which command would give us that info. I think that once we move reachaibility to versioning there will be no issue, but until we do, it makes sense to give it priority in the short/long description.

File: temporalcli/commandsmd/commands.md
Comment (antlai-temporal) [inline temporalcli/commandsmd/commands.md:769]: for reachability we use `report-reachability` as opposed to `disable-reachability` because it is expensive, and we wanted to avoid calls.
Is getting the stats also expensive? In that case we should use also use `report-stats`...

Comment (ShahabT) [reply temporalcli/commandsmd/commands.md:769]: Stats is not the cheapest thing (because it need partition fan-out) but still relatively cheap. I'm comfortable with having it enabled by default. wdyt @dnr ?

We may want to add caching for stats in the root partition in the future (if SDK starts to query it for poller scaling).

File: temporalcli/commandsmd/commands.md
Comment (ShahabT) [inline temporalcli/commandsmd/commands.md:728]: this part of the sentence is only about reachability and it should be removed.

File: temporalcli/commandsmd/commands.md
Comment (ShahabT) [inline temporalcli/commandsmd/commands.md:754]:
```suggestion
Task reachability information is returned for the requested versions and all task types, which can be used to safely retire Workers with old code versions, provided that they were assigned a Build ID.

Note that task reachability status is experimental and may significantly change or be removed in a future release.

The reachability states of a Build ID are:
```

File: temporalcli/commandsmd/commands.md
Comment (Shivs11) [reply temporalcli/commandsmd/commands.md:728]: I agree, however we should have the part : `the requested versions and all task types`. 

All in all, the sentence should look like:
The `temporal task-queue describe` command provides poller information, backlog statistics and task reachability (experimental) information for the requested versions and task types.

Comment (ShahabT) [reply temporalcli/commandsmd/commands.md]: > task reachability (experimental) ~information~

"information" seems redundant?

**Review by ShahabT — APPROVED** [review]

**Pull request merged**
